lix/packages/sdk/src/branch/get-leaf-changes-diff.ts

```diff
@@ -0,0 +1,113 @@
+import type { Change, LixReadonly } from "@lix-js/sdk";
+import type { LixPlugin } from "../plugin.js";
+
+export async function getLeafChangesDiff(args: {
+	plugins: LixPlugin[];
+	sourceDb: LixReadonly["db"];
+	targetDb?: LixReadonly["db"];
+	sourceBranchId?: string;
+	targetBranchId?: string;
+}): Promise<Change[]> {
+	if (!args.targetDb) {
+		args.targetDb = args.sourceDb;
+	}
+	if (!args.sourceBranchId) {
+		args.sourceBranchId = (
+			await args.sourceDb
+				.selectFrom("branch")
+				.select("id")
+				.where("active", "=", true)
+				.executeTakeFirstOrThrow()
+		).id;
+	}
+	if (!args.targetBranchId) {
+		args.targetBranchId = (
+			await args.targetDb
+				.selectFrom("branch")
+				.select("id")
+				.where("active", "=", true)
+				.executeTakeFirstOrThrow()
+		).id;
+	}
+
+	const leafChangesInSource = await args.sourceDb
+		.selectFrom("change_view")
+		.selectAll()
+		.where("branch_id", "=", args.sourceBranchId)
+		.where(
+			"id",
+			"not in",
+			// @ts-ignore - no idea what the type issue is
+			args.sourceDb
+				.selectFrom("change_view")
+				.select("parent_id")
+				.where("parent_id", "is not", undefined)
+				.where("branch_id", "=", args.sourceBranchId)
+				.distinct(),
+		)
+		.execute();
+
+	const sourceChangesByValueId = leafChangesInSource.reduce((agg, change) => {
+		// @ts-ignore
+		agg[change.value.id] = change;
+		return agg;
+	}, {});
+
+	const leafChangesInTarget = await args.targetDb
+		.selectFrom("change_view")
+		.selectAll()
+		.where("branch_id", "=", args.targetBranchId)
+		.where(
+			"id",
+			"not in",
+			// @ts-ignore - no idea what the type issue is
+			args.targetDb
+				.selectFrom("change_view")
+				.select("parent_id")
+				.where("parent_id", "is not", undefined)
+				.where("branch_id", "=", args.targetBranchId)
+				.distinct(),
+		)
+		.execute();
+
+	const targetChangesByValueId = leafChangesInTarget.reduce((agg, change) => {
+		// @ts-ignore
+		agg[change.value.id] = change;
+		return agg;
+	}, {});
+
+	const toDelete = [];
+	for (const change of leafChangesInTarget) {
+		// @ts-ignore
+		if (!sourceChangesByValueId[change.value.id]) {
```

Comment:
Wouldn't this delete an entry that was created in the target and does not exist in the source - while not really been deleted in source?

Reply:
@martin-lysk  yes that is the goal here. remember this does not delete the change entry just the effect it has on the file table which is exactly what switching from source to target branch means. ping me if this does not make sense and we need to discuss...